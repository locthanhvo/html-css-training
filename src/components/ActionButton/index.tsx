import { JSXElementConstructor, ReactElement, memo } from 'react';
import { Button } from '@chakra-ui/react';

interface ActionButtonProps {
  type?: 'button' | 'submit' | 'reset';
  size?: 'sm' | 'md' | 'lg' | 'xl';
  isLoading?: boolean;
  isDisabled?: boolean;
  title: string;
  bgColor?: string;
  color?: string;
  borderColor?: string;
  onClick?: () => void;
  rightIcon?:
    | ReactElement<any, string | JSXElementConstructor<any>>
    | undefined;
}

const ActionButton = ({
  title,
  type = 'button',
  size,
  isLoading,
  isDisabled,
  bgColor,
  color,
  borderColor,
  rightIcon,
  onClick,
}: ActionButtonProps) => {
  return (
    <Button
      type={type}
      name={title}
      size={size}
      bgColor={bgColor}
      color={color}
      border="1px solid"
      borderColor={borderColor}
      onClick={onClick}
      rightIcon={rightIcon}
      isLoading={isLoading}
      isDisabled={isDisabled}
    >
      {title}
    </Button>
  );
};

export default memo(ActionButton);
