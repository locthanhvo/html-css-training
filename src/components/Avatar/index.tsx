import { Center, Flex, Image, Text } from '@chakra-ui/react';

import AvatarDefault from '../../assets/images/avatar.png';
import { memo } from 'react';

interface AvatarProps {
  url?: string;
  firstName: string;
  lastName: string;
  email: string;
}

const Avatar = ({ url, firstName, lastName, email }: AvatarProps) => {
  return (
    <Flex gap={2} alignItems="center">
      <Center w={10} h={10} borderRadius="50%">
        <Image src={url || AvatarDefault} />
      </Center>

      <Flex direction="column">
        <Text color="primary" fontWeight="medium" fontSize="base">
          {firstName} {lastName}
        </Text>
        <Text color="primary" fontWeight="regular" fontSize="xs">
          {email}
        </Text>
      </Flex>
    </Flex>
  );
};

export default memo(Avatar);
