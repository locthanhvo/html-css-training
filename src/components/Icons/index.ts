export * from './AddIcon';
export * from './ForwardIcon';
export * from './SearchIcon';
export * from './ArrowIcon';
export * from './BrandIcon';
export * from './FolderIcon';
export * from './ProjectIcon';
export * from './DirectionIcon';
export * from './ProtectionIcon';
export * from './ConfigIcon';
export * from './BreadcrumbIcon';
export * from './NotifyIcon';
export * from './DropdownIcon';
export * from './DeleteIcon';
export * from './EditIcon';
export * from './LogoIcon';
