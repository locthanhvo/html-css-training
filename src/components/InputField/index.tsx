import {
  ChangeEvent,
  ReactNode,
  memo,
  useCallback,
  forwardRef,
  ForwardedRef,
} from 'react';
import {
  Input,
  InputGroup,
  InputLeftElement,
  InputProps,
  FormControl,
  FormLabel,
  FormErrorMessage,
} from '@chakra-ui/react';

type TInputFieldProps = Omit<InputProps, 'onChange'> & {
  type?: 'text' | 'datetime-local' | 'email' | 'password';
  isValidate?: boolean;
  isError?: boolean;
  errorMessages?: string;
  label?: string;
  leftIcon?: ReactNode;
  onChange: (value: string) => void;
};

const InputField = (
  {
    isError = false,
    errorMessages = 'Default error',
    type = 'text',
    label,
    leftIcon,
    onChange,
    ...rest
  }: TInputFieldProps,
  ref: ForwardedRef<HTMLInputElement>,
) => {
  const handleChangeValue = useCallback(
    (e: ChangeEvent<HTMLInputElement>): void => onChange(e.target.value),
    [onChange],
  );

  return (
    <FormControl isInvalid={isError}>
      {label && (
        <FormLabel
          fontSize="base"
          fontWeight="regular"
          marginInlineEnd={0}
          minW="max-content"
        >
          {label}
        </FormLabel>
      )}
      <InputGroup>
        {leftIcon && (
          <InputLeftElement pointerEvents="none">{leftIcon}</InputLeftElement>
        )}

        <Input
          py={5}
          type={type}
          color="text.primary"
          onChange={handleChangeValue}
          ref={ref}
          {...rest}
          isInvalid={isError}
          _focus={{ border: '1px solid', borderColor: 'blue.100' }}
        />
      </InputGroup>
      {isError && (
        <FormErrorMessage fontSize="2xs">{errorMessages}</FormErrorMessage>
      )}
    </FormControl>
  );
};

export default memo(forwardRef(InputField));
