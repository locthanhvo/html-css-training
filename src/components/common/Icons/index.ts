export { default as UserProfileIcon } from './UserProfileIcon';
export { default as BuildingIcon } from './BuildingIcon';
export { default as WorldIcon } from './WorldIcon';
export { default as ForwardArrowIcon } from './ForwardArrowIcon';
export { default as PlusCircleIcon } from './PlusCircleIcon';
export { default as CompassIcon } from './CompassIcon';
export { default as FavoriteIcon } from './FavoriteIcon';
export { default as MenuIcon } from './MenuIcon';
export { default as MessageIcon } from './MessageIcon';
export { default as NotifyIcon } from './NotifyIcon';
export { default as SearchIcon } from './SearchIcon';
export { default as StarIcon } from './StarIcon';
export { default as StatisticsIcon } from './StatisticsIcon';
export { default as MeatballsMenuIcon } from './MeatballsMenuIcon';
export { default as BarChartIcon } from './BarChartIcon';
export { default as ChecklistIcon } from './ChecklistIcon';
export { default as CloudIcon } from './CloudIcon';
export { default as EmailIcon } from './EmailIcon';
export { default as UserGroupIcon } from './UserGroupIcon';
export { default as CalendarIcon } from './CalendarIcon';
export { default as SettingIcon } from './SettingIcon';
export { default as PhoneIcon } from './PhoneIcon';
export { default as BookIcon } from './BookIcon';
export { default as TelegramIcon } from './TelegramIcon';
export { default as WhatsappIcon } from './WhatsappIcon';
export { default as ClockIcon } from './ClockIcon';
export { default as AttachmentIcon } from './AttachmentIcon';
export { default as HamburgerMenuIcon } from './HamburgerMenuIcon';
export { default as ProfileIcon } from './ProfileIcon';
export { default as PlusIcon } from './PlusIcon';
export { default as BurgerMenuLeftIcon } from './BurgerMenuLeftIcon';
export { default as FlagIcon } from './FlagIcon';
export { default as CloseIcon } from './CloseIcon';
