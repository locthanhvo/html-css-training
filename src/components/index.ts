export { default as ActionButton } from './ActionButton';
export { default as Status } from './Status';
export { default as InputField } from './InputField';
export { default as Header } from './Header';
export { default as NavTab } from './NavTab';
export { default as Avatar } from './Avatar';
export { default as Modal } from './Modal';
export { default as ExpandSidebar } from './ExpandSidebar';
export { default as Table } from './Table';
export { default as ErrorBoundary } from './ErrorBoundary';
export { default as Fallback } from './Fallback';
export { default as Dropdown } from './Dropdown';
export { default as Pagination } from './Pagination';
export * from './Form';
