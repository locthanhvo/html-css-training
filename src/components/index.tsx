export { default as NavTab } from './NavTab';
export { default as Sidebar } from './SideBar';
export { default as Header } from './Header';
export { default as CardList } from './CardList';
