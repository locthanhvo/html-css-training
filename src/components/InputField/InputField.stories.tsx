import { StoryObj, Meta } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import InputField from '.';

import { SearchIcon } from '../Icons';

const defaultProps = {
  name: 'email',
  placeholder: 'Email',
  onChange: () => {},
};

const meta: Meta<typeof InputField> = {
  title: 'Components/InputField',
  tags: ['autodocs'],
  component: InputField,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  argTypes: {
    name: {
      description: 'The name of the input field',
      defaultValue: 'email',
    },

    placeholder: {
      description: 'The placeholder for the input field',
      defaultValue: 'email',
    },

    isValidate: {
      description:
        'The validation state of the input field (the default state is false)',
    },

    isError: {
      description:
        'The error state of the input field (the default state is false)',
    },

    errorMessages: {
      description: 'The error messages of the input field',
      defaultValue: 'Default error',
    },

    label: {
      description: 'The label for the input field',
    },

    leftIcon: {
      description: 'The icon on the right of the input field',
    },

    onChange: {
      description:
        'The onChange function that handles the change of the input field',
      action: 'onChange',
    },
  },
  parameters: {
    controls: {
      expanded: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof InputField>;

export const Default: Story = {
  args: {
    ...defaultProps,
  },
};

export const HasLeftIcon: Story = {
  args: {
    ...defaultProps,
    leftIcon: <SearchIcon />,
  },
};

export const HasError: Story = {
  args: {
    ...defaultProps,
    isValidate: true,
    isError: true,
  },
};

export const DatePicker: Story = {
  args: {
    ...defaultProps,
    type: 'datetime-local',
  },
};
