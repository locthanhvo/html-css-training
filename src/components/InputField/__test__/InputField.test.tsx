import { render } from '@testing-library/react';

// Components
import InputField from '..';

// Icons
import { SearchIcon } from '@/components/Icons';

describe('InputField component', () => {
  const mockOnChange = jest.fn();
  const props = {
    name: 'email',
    placeholder: 'Email',
    onChange: mockOnChange,
  };
  it('Render correctly with default props', () => {
    const { container } = render(<InputField {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('Render correctly with right icon', () => {
    const { container } = render(
      <InputField {...props} leftIcon={<SearchIcon />} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('Render correctly with error message', () => {
    const { container } = render(
      <InputField
        {...props}
        isError
        isValidate
        errorMessages="Field is required"
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
