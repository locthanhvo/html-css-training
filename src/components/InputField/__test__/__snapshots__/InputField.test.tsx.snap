// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputField component Render correctly with default props 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    name="email"
    placeholder="Email"
    role="group"
  >
    <div
      class="chakra-input__group css-bx0blc"
      data-group="true"
    >
      <input
        class="chakra-input css-adz8kx"
        id="field-:r0:"
        name="email"
        placeholder="Email"
        type="text"
      />
    </div>
  </div>
</div>
`;

exports[`InputField component Render correctly with error message 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    data-invalid=""
    name="email"
    placeholder="Email"
    role="group"
  >
    <div
      class="chakra-input__group css-bx0blc"
      data-group="true"
    >
      <input
        aria-describedby="field-:r2:-feedback"
        aria-invalid="true"
        class="chakra-input css-adz8kx"
        id="field-:r2:"
        name="email"
        placeholder="Email"
        type="text"
      />
    </div>
    <div
      aria-live="polite"
      class="chakra-form__error-message css-1qc3jta"
      id="field-:r2:-feedback"
    >
      Field is required
    </div>
  </div>
</div>
`;

exports[`InputField component Render correctly with right icon 1`] = `
<div>
  <div
    class="chakra-form-control css-0"
    name="email"
    placeholder="Email"
    role="group"
  >
    <div
      class="chakra-input__group css-bx0blc"
      data-group="true"
    >
      <div
        class="chakra-input__left-element css-1u2k0bl"
      >
        <svg
          fill="none"
          height="18"
          viewBox="0 0 18 18"
          width="18"
        >
          <path
            d="M8.1111 15.2222C12.0385 15.2222 15.2222 12.0385 15.2222 8.1111C15.2222 4.18375 12.0385 1 8.1111 1C4.18375 1 1 4.18375 1 8.1111C1 12.0385 4.18375 15.2222 8.1111 15.2222Z"
            stroke="#18181B"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M17 16.9998L13.1333 13.1332"
            stroke="#18181B"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <input
        class="chakra-input css-adz8kx"
        id="field-:r1:"
        name="email"
        placeholder="Email"
        type="text"
      />
    </div>
  </div>
</div>
`;
